{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "### standard python pacakges\n",
    "\n",
    "import os\n",
    "import glob\n",
    "import re\n",
    "import csv\n",
    "import subprocess\n",
    "\n",
    "### custom python\n",
    "\n",
    "from generate_summary_csv import sample_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_dict = {'H4':'ko3','A5':'ko4','F3':'ko1','G3':'ctl1','H3':'ko2','A4':'ctl2'}\n",
    "\n",
    "# sample_dict = {'H4':'ko3'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "chemistry_dict = {'ko3':'v3','ko4':'v3','ko1':'v3','ctl1':'v3','ko2':'v3','ctl2':'v3'}\n",
    "\n",
    "# chemistry_dict = {'ko3':'v3'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "parameter_dict = {'transcriptome_path':'XXX',\n",
    "                  'barcode_rank_threshold':500,\n",
    "                  'droplet_utils_FDR':0.01,\n",
    "                  'mito_threshold':20}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "class sequence_project:\n",
    "    def __init__(self, directory_path, project_name):\n",
    "        self.directory_path = directory_path\n",
    "        self.project_name = project_name\n",
    "        self.sample_dict = sample_dict\n",
    "        self.chemistry_dict = chemistry_dict\n",
    "        self.sub_dict = {}\n",
    "        self.parameter_dict = parameter_dict\n",
    "        self.output_directory = self.directory_path.rsplit('/',1)[0] + '/' + self.project_name\n",
    "        self.file_list = self.get_file_list()\n",
    "        self.file_dict = {}\n",
    "        self.file_dict\n",
    "        self.generate_file_dict()\n",
    "        self.date_lane_dict = {}\n",
    "        self.generate_date_to_lane_dict()\n",
    "        self.run_the_pipeline() \n",
    "    \n",
    "    def get_file_list(self):\n",
    "        glob_input = self.directory_path + '/*.fastq.gz'\n",
    "        file_list = glob.glob(glob_input)\n",
    "        return file_list\n",
    "    \n",
    "    def generate_file_dict(self):\n",
    "        for key in sample_dict:\n",
    "            search_string = '-' + key + '-'\n",
    "            relevant_files = list(filter(lambda x: search_string in x, self.file_list))\n",
    "            self.file_dict[key] = relevant_files\n",
    "    \n",
    "    def generate_date_to_lane_dict(self):\n",
    "        for file in self.file_list:\n",
    "            date = re.search('_......_', file).group(0)[1:-1]\n",
    "            self.date_lane_dict[date] = \"\"\n",
    "        lane_num = 0\n",
    "        for key in self.date_lane_dict:\n",
    "            lane_num = lane_num + 1\n",
    "            lane_string = 'L0'\n",
    "            if lane_num < 10:\n",
    "                lane_string = lane_string + '0' + str(lane_num)\n",
    "            else:\n",
    "                lane_string = lane_string + str(lane_num)\n",
    "            self.date_lane_dict[key] = lane_string\n",
    "        \n",
    "    def make_output_dir(self):\n",
    "        os.system('mkdir '+ self.output_directory)\n",
    "        os.chdir(self.output_directory)     \n",
    "    \n",
    "    def rename_and_move_files(self):\n",
    "        old_new_tuple = []\n",
    "        x= 0\n",
    "        for key in self.file_dict:\n",
    "            old_path_list = self.file_dict[key]\n",
    "            for path in old_path_list:\n",
    "                old_file_path = path\n",
    "                old_file = path.split('/')[-1]\n",
    "                old_path = path.rsplit('/',1)[0] + '/'\n",
    "                sample_name = re.search('GA-..-', old_file).group(0)[3:-1]\n",
    "                date = re.search('_......_', old_file).group(0)[1:-1]\n",
    "                sample_string = self.sub_dict[sample_name]\n",
    "                print('1')\n",
    "                print(date)\n",
    "                print(sample_string)\n",
    "                lane_string = '_' + 'L001' + \"_\" #self.date_lane_dict[date] + \"_\"\n",
    "                print('2')\n",
    "                read_string = re.search('_.._', old_file).group(0)[1:-1] + '_'\n",
    "                end_string = '001.fastq.gz'\n",
    "                new_file_name = sample_string + '_S1' + lane_string + read_string + end_string\n",
    "                new_file_path = self.output_directory + '/' + new_file_name\n",
    "                old_new_tuple.append(tuple((old_file_path, new_file_path)))\n",
    "        for name in old_new_tuple:\n",
    "            command_list = ['cp',str(name[0]),str(name[1])]\n",
    "            os_copy_and_move_command = \" \".join(command_list)\n",
    "            os.system(os_copy_and_move_command)\n",
    "    \n",
    "    def run_cellranger(self):\n",
    "        for key in self.sub_dict:\n",
    "            id_name = self.sub_dict[key]\n",
    "            transcriptome = self.parameter_dict['transcriptome_path']\n",
    "            fastqs = self.output_directory\n",
    "            sample = self.sub_dict[key]\n",
    "            cellranger_command = \"cellranger count --id={} --include-introns --transcriptome={} --fastqs={} --sample={}\".format(id_name,\\\n",
    "                                                                                                         transcriptome,\\\n",
    "                                                                                                         fastqs,\\\n",
    "                                                                                                         sample)\n",
    "            print(cellranger_command)\n",
    "            os.system(cellranger_command)\n",
    "            \n",
    "            \n",
    "\n",
    "    def cellranger_csv_summary(self):\n",
    "        csv_summary_paths = []\n",
    "        sample_names = self.sub_dict.values()\n",
    "        for name in sample_names:\n",
    "            new_path = self.output_directory + '/' + name +'/outs/metrics_summary.csv'\n",
    "            csv_summary_paths.append(new_path)\n",
    "        output_location = self.output_directory + '/' + self.project_name + '_cellranger_summary.csv'\n",
    "        sample_summary(csv_summary_paths, output_location,self.chemistry_dict)\n",
    "        subprocess.call(['Rscript', 'cellranger_summary_graph.r', output_location])   ### generate a graph in R\n",
    "    \n",
    "    def pre_process(self):\n",
    "        sample_names = self.sub_dict.values()\n",
    "        project_directory = self.output_directory\n",
    "        arg_2 = str(self.parameter_dict['barcode_rank_threshold'])\n",
    "        arg_3 = str(self.parameter_dict['droplet_utils_FDR'])\n",
    "        arg_4 = str(self.parameter_dict['mito_threshold'])\n",
    "        arg_6 = '/'.join([self.output_directory,'pre_process_objects'])\n",
    "        arg_7 = project_directory\n",
    "        os.system('mkdir '+ arg_6)\n",
    "        for sample in sample_names:\n",
    "            sample_directory = '/'.join([self.output_directory,sample,'outs/raw_feature_bc_matrix'])\n",
    "            arg_1 = sample_directory\n",
    "            arg_5 = sample\n",
    "            subprocess.call(['Rscript', 'pre_process.r', arg_1, arg_2, arg_3, arg_4, arg_5, arg_6, arg_7])\n",
    "    \n",
    "    def merge_and_align(self):\n",
    "        arg_1 = '/'.join([self.output_directory,'pre_process_objects'])\n",
    "        arg_2 = self.output_directory\n",
    "        subprocess.call(['Rscript','merge_and_align.r',arg_1,arg_2])\n",
    "\n",
    "    def run_the_pipeline(self):\n",
    "        sample_list = [sample for sample in dict.values(self.sample_dict)]\n",
    "        number_list = [number for number in range(len(sample_list))]\n",
    "        barcode_list = [barcode for barcode in dict.keys(self.sample_dict)]\n",
    "        sample_index_list = list(zip(sample_list,barcode_list,number_list))\n",
    "        print(\"\"\"Hello user. Here are the samples available for processing today: \\n\"\"\")\n",
    "        print(\"\"\"Sample\\tBarcode\\tIndex\"\"\")\n",
    "        for sample in sample_index_list:\n",
    "            print('{}:\\t{}:\\t{}'.format(sample[0],sample[1],sample[2]))\n",
    "        user_index = input(\"\\n\" \\\n",
    "        \"What samples would you like to process? Please enter one of the following: \\n\" \\\n",
    "        \"a) a comma separated list of indices\\n\" \\\n",
    "        \"b) all, to process all samples\\n\" \\\n",
    "        \"\\n\")\n",
    "\n",
    "        if user_index == 'all':\n",
    "            self.sub_dict = self.sample_dict\n",
    "        else:\n",
    "            index_list = user_index.split(',')\n",
    "            index_list = [int(x) for x in index_list]\n",
    "            for index in index_list:\n",
    "                barcode = barcode_list[index]\n",
    "                sample = sample_list[index]\n",
    "                self.sub_dict[barcode] = sample\n",
    "\n",
    "        process_select = input(\"\\n\" \\\n",
    "                               \"Terrific, where should we start (enter a #)?:\\n\" \\\n",
    "                               \"1 = start from scratch\\n\" \\\n",
    "                               \"2 = start from aligned cell-ranger input\\n\"  \\\n",
    "                               \"3 = start from seurat processed input\\n\" \\\n",
    "                               \"\\n\")\n",
    "\n",
    "        print(\"\\n\")\n",
    "\n",
    "        start_value = int(process_select)\n",
    "        if start_value == 1:\n",
    "            self.make_output_dir()\n",
    "            self.rename_and_move_files()\n",
    "            self.run_cellranger()\n",
    "            #print(self.file_dict)\n",
    "            #self.cellranger_csv_summary()\n",
    "            print(self.file_list)\n",
    "            print(self.directory_path)\n",
    "            start_value +=1\n",
    "        if start_value == 2:\n",
    "            start_value +=1\n",
    "            #self.pre_process()\n",
    "        if start_value == 3:\n",
    "         print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nr1_prelim = sequence_project('XXX','nr1_all_processed_flx_expand')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
